### PR TITLE
Add support for InfiniBand

### PR DESCRIPTION
**Pull request by bdrung**
The Prometheus node exporter collects metrics for the InfiniBand network protocol including the amount of packets sent and received, the number of times the link has been downed and how many times the link has recovered from an error state.

Reading all those information from sysfs is better placed in the procfs library. Also collect the state, physical state, and the rate.

### CODE REVIEW

**Comment by bdrung** [comment]
Thanks for the review. I will address those points. @mdlayher, do you really prefer to have a switch statement for the 16 different values? This list might grow. If the counter fields are regular values, what should I do if the system does not have this counter?

File: sysfs/class_infiniband.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+		res = append(res, deviceDir.Name())
+	}
+
+	return res, nil
+}
+
+// InfiniBandDevicePorts scans /sys/class/infiniband/<device> for ports and returns them as a list of integers.
+func (fs FS) InfiniBandDevicePorts(device string) ([]uint, error) {
+	var res []uint
+	path := fs.sys.Path(filepath.Join(infinibandclassPath, device, "ports"))
```
[the remark this answers is not in the document]
Comment (bdrung) [reply sysfs/class_infiniband.go]: Done.

File: sysfs/class_infiniband.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+	infiniBandClass := InfiniBandClass{}
+	for _, device := range devices {
+		ports, err := fs.InfiniBandDevicePorts(device)
+		if err != nil {
+			return nil, err
+		}
+
+		infiniBandClass[device] = map[uint]InfiniBandPort{}
+		for _, port := range ports {
+			path := fs.sys.Path(filepath.Join(infinibandclassPath, device, "ports", fmt.Sprint(port)))
```
[the remark this answers is not in the document]
Comment (bdrung) [reply sysfs/class_infiniband.go]: Done.

File: sysfs/class_infiniband.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+		if err != nil {
+			return res, fmt.Errorf("failed to convert %s into uint", portDir.Name())
+		}
+		res = append(res, uint(port))
+	}
+
+	return res, nil
+}
+
+// NewNetClass returns info for all ports (port) of all InfiniBand devices (device) read from /sys/class/infiniband/<device>/ports/<port>.
```
[the remark this answers is not in the document]
Comment (bdrung) [reply sysfs/class_infiniband.go]: Done.

File: sysfs/class_infiniband.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+func (fs FS) NewInfiniBandClass() (InfiniBandClass, error) {
```
[the remark this answers is not in the document]
Comment (bdrung) [reply sysfs/class_infiniband.go]: Okay. Renamed to `InfiniBandClass`.

File: sysfs/class_infiniband.go [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+	if err != nil {
+		return rate, fmt.Errorf("failed to convert %s into uint", strings.TrimSpace(s[0]))
+	}
+	rate = value * 125000000
+	return rate, nil
+}
+
+// parseInfiniBandPort scans predefined files in /sys/class/infiniband/<device>/ports/<port>
+// directory and gets their contents.
+func (_ InfiniBandClass) parseInfiniBandPort(portPath string) (*InfiniBandPort, error) {
```
[the remark this answers is not in the document]
Comment (bdrung) [reply sysfs/class_infiniband.go]: Done.

**Comment by bdrung** [comment]
I looked at https://github.com/prometheus/node_exporter/blob/master/collector/infiniband_linux.go and borrowed the workarounds and the legacy counters.

**Comment by bdrung** [comment]
I changed the two parse functions to take a string as parameter (instead of a path to the file).

File: sysfs/class_infiniband.go [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+		return id, name, fmt.Errorf("failed to convert %s into uint", strings.TrimSpace(parts[0]))
+	}
+	id = uint(value)
+	return id, name, nil
+}
+
+// Parse rate (example: "100 Gb/sec (4X EDR)") and return it as bytes/second
+func parseRate(s string) (uint64, error) {
+	var rate uint64
+	parts := strings.Split(s, "Gb/sec")
```
Comment (discordianfish) [inline sysfs/class_infiniband.go]: Is this the only unit possible?

**Comment by discordianfish** [comment]
Tend to agree with @mdlayher. What about using a temporary map to iterate over instead of using reflect to iterate over the map?

File: sysfs/class_infiniband.go (hunk 6 above)
Comment (dswarbrick) [reply sysfs/class_infiniband.go]: Yes - see `rate_show` function in https://elixir.bootlin.com/linux/v5.1.7/source/drivers/infiniband/core/sysfs.c:
```c
	return sprintf(buf, "%d%s Gb/sec (%dX%s)\n",
		       rate / 10, rate % 10 ? ".5" : "",
		       ib_width_enum_to_int(attr.active_width), speed);
```

Comment (bdrung) [reply sysfs/class_infiniband.go]: I changed to code to also correctly parse `2.5 Gb/sec` and added a test case for it.

**Comment by bdrung** [comment]
`sysfs/class_power_supply.go` and `sysfs/net_class.go` uses reflect. Can you point me to a component that uses a temporary map to iterate over?

**Comment by bdrung** [comment]
@mdlayher I saw your recent commits on master and adapted this pull request to the logic used for PowerSupply. Therefore `reflect` is not used any more. Additional changes:
* Introduce `InfiniBandDevice` type
* also collect `board_id`, `fw_ver`, and `hca_type` (in InfiniBandDevice)

All remarks are addressed and this pull request is ready for an in depth review.

File: internal/util/valueparser.go [hunk 7]
```diff
@@ -52,6 +52,25 @@ func (vp *ValueParser) PInt64() *int64 {
 	return &v
 }
 
+// PUInt64 interprets the underlying value as an uint64 and returns a pointer to
+// that value.
+func (vp *ValueParser) PUInt64() *uint64 {
```
Comment (mdlayher) [inline internal/util/valueparser.go]: Mind adding a test for this new method as well? You can probably just duplicate the PInt64 cases.

File: sysfs/class_infiniband.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+
+	return ibc, nil
+}
+
+// Parse one InfiniBand device.
+func (fs FS) parseInfiniBandDevice(name string) (*InfiniBandDevice, error) {
+	path := fs.sys.Path(infinibandClassPath, name)
+	device := InfiniBandDevice{Name: name}
+
+	for _, f := range [3]string{"board_id", "fw_ver", "hca_type"} {
```
Comment (mdlayher) [inline sysfs/class_infiniband.go]: Nit: how about `[...]` so the size will be inferred automatically? I'd usually just go with a slice but an array seems reasonable since this will probably never change.

File: sysfs/class_infiniband.go [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+
+		device.Ports[port.Port] = *port
+	}
+
+	return &device, nil
+}
+
+// Parse InfiniBand state. Expected format: "<id>: <string-representation>"
+func parseState(s string) (uint, string, error) {
+	var id uint
```
Comment (mdlayher) [inline sysfs/class_infiniband.go]: IMO remove these and just use the usual `:=` syntax below.

File: sysfs/class_infiniband.go [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+	var rate uint64
+	parts := strings.Split(s, "Gb/sec")
+	if len(parts) != 2 {
+		return rate, fmt.Errorf("failed to split '%s' by 'Gb/sec'", s)
+	}
+	value, err := strconv.ParseFloat(strings.TrimSpace(parts[0]), 32)
+	if err != nil {
+		return rate, fmt.Errorf("failed to convert %s into uint", strings.TrimSpace(parts[0]))
+	}
+	rate = uint64(value * 125000000)
```
Comment (mdlayher) [inline sysfs/class_infiniband.go]: Same, `:=`.

Can you leave a comment about the meaning of this constant?

File: sysfs/class_infiniband.go [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+}
+
+// parseInfiniBandPort scans predefined files in /sys/class/infiniband/<device>/ports/<port>
+// directory and gets their contents.
+func (fs FS) parseInfiniBandPort(name string, port string) (*InfiniBandPort, error) {
+	portNumber, err := strconv.ParseUint(port, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert %s into uint", port)
+	}
+	infiniBandPort := InfiniBandPort{Name: name, Port: uint(portNumber)}
```
Comment (mdlayher) [inline sysfs/class_infiniband.go]: Nit: you can go with a shorter name since this is confined to a single function. Maybe `ibp`?

File: sysfs/class_infiniband.go [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+		name := filepath.Join(path, f.Name())
+		value, err := util.SysReadFile(name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file %q: %v", name, err)
+		}
+
+		// According to Mellanox, the metrics port_rcv_data, port_xmit_data,
+		// port_rcv_data_64, and port_xmit_data_64 "are divided by 4 unconditionally"
+		// as they represent the amount of data being transmitted and received per lane.
+		// Mellanox cards have 4 lanes per port, so all values must be multiplied by 4
```
Comment (mdlayher) [inline sysfs/class_infiniband.go]: Does anyone other than Mellanox make infiniband gear?

File: sysfs/class_infiniband.go [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+)
+
+const infinibandClassPath = "class/infiniband"
+
+// InfiniBandCounters contains counter values from files in
+// /sys/class/infiniband/<Name>/ports/<Port>/counters or
+// /sys/class/infiniband/<Name>/ports/<Port>/counters_ext
+// for a single port of one InfiniBand device.
+type InfiniBandCounters struct {
+	LegacyPortMulticastRcvPackets  *uint64 // counters_ext/port_multicast_rcv_packets
```
Comment (mdlayher) [inline sysfs/class_infiniband.go]: Do we have to have the legacy versions as well? Would it be possible to look for both new/legacy and only use legacy if the new version is absent?

File: sysfs/class_infiniband_test.go [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+		t.Fatal(err)
+	}
+
+	got, err := fs.InfiniBandClass()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var (
+		port1LinkDowned               uint64 = 0
```
Comment (mdlayher) [inline sysfs/class_infiniband_test.go]: Nit: omit any `= 0`; it's inferred automatically.

File: sysfs/class_infiniband_test.go [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+package sysfs
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestParseSlowRate(t *testing.T) {
```
Comment (mdlayher) [inline sysfs/class_infiniband_test.go]: Nit: combine this and the below test into a table-driven test.

File: sysfs/class_infiniband.go (hunk 12 above)
Comment (dswarbrick) [reply sysfs/class_infiniband.go]: Mellanox (recently acquired by Nvidia) dominate the IB market by far, but previously Qlogic (tech acquired by Intel) and Voltaire (merged with Mellanox) produced IB hardware. Oracle produces their own IB chipsets.

File: sysfs/class_infiniband.go (hunk 13 above)
Comment (dswarbrick) [reply sysfs/class_infiniband.go]: We run a mixture of kernel versions where I work, and I've had to do some elaborate metric relabelling to make our IB metric names conform to a common standard, so that they're supported by a one-size-fits-all Grafana dashboard. Brace yourself for a verbose reason why...

Kernels 4.5 and later merge the `counters_ext` IB counters in sysfs into the `counters` directory. This effectively replaces the 32-bit counters that were previously exposed in the `counters` directory with the 64-bit counters from the `counters_ext` directory.

Unfortunately, InfiniBand counters do not automatically roll over, so when the 32-bit counters reach 4 billion, they simply latch at that value (which is obviously bad for `rate` / `irate` queries).

Node_exporter doesn't really handle this in a sane way, so on pre-4.5 kernels it exposes 32-bit counters as e.g. `node_infiniband_port_packets_transmitted_total` and 64-bit counters as
e.g. `node_infiniband_legacy_packets_transmitted_total`. This makes it difficult to have unified dashboards / alerting rules which can handle both pre- and post-4.5 kernels.

To cut a long story short, I ended up configuring the following relabelling rules, which rename the legacy metrics if exposed, and overwrite the "non-legacy" metrics (which will be 32-bit, and thus useless, if legacy metrics are present):
```yaml
  # Rename node_infiniband_legacy_* (pre-MLNX_OFED 4.0) to unify PromQL queries
  - source_labels: [__name__]
    target_label: __name__
    regex: node_infiniband_legacy_((?:data|packets)_.+)
    replacement: node_infiniband_port_$1
  - source_labels: [__name__]
    target_label: __name__
    regex: node_infiniband_legacy_((?:multicast|unicast)_.+)
    replacement: node_infiniband_$1
```
In short - it would be nice if node_exporter simply exposed the widest counters available, regardless of kernel / OFED version. However, that is something that needs to happen in node_exporter, not procfs, and it may be desirable to still expose both legacy (32-bit) and modern (64-bit) counters, at least in procfs.

File: internal/util/valueparser.go (hunk 7 above)
Comment (bdrung) [reply internal/util/valueparser.go]: Done.

File: sysfs/class_infiniband.go (hunk 8 above)
Comment (bdrung) [reply sysfs/class_infiniband.go]: Done.

File: sysfs/class_infiniband.go (hunk 10 above)
Comment (bdrung) [reply sysfs/class_infiniband.go]: Done. Added: "Convert Gb/s into bytes/s"

File: sysfs/class_infiniband_test.go (hunk 14 above)
Comment (bdrung) [reply sysfs/class_infiniband_test.go]: Done, but I like to prefer the explicit definition.

File: sysfs/class_infiniband_test.go (hunk 15 above)
Comment (bdrung) [reply sysfs/class_infiniband_test.go]: Done.

File: sysfs/class_infiniband.go (hunk 11 above)
Comment (bdrung) [reply sysfs/class_infiniband.go]: Done.

File: sysfs/class_infiniband.go (hunk 13 above)
Comment (bdrung) [reply sysfs/class_infiniband.go]: Thanks for the answer @dswarbrick. Therefore I mark this conversation as resolved without changing the code.

**Comment by pgier** [comment]
Merged, thanks!
